src/core/namespace.c

```diff
@@ -287,10 +288,18 @@ static int apply_mount(
 
                 /* First, get rid of everything that is below if there
                  * is anything... Then, overmount it with an
-                 * inaccessible directory. */
+                 * inaccessible path. */
                 umount_recursive(m->path, 0);
 
-                what = "/run/systemd/inaccessible";
+                // mount() always follow symlinks, so we follow symlinks during the stat() as well
+                r = stat(m->path, &target);
+                if (r != 0)
+                        return r;
+                if (!S_ISDIR(target.st_mode))
+                        what = "/dev/null";
+                else {
+                        what = "/run/systemd/inaccessible";
+                }
                 break;
```

Comment:
so, as far as i could verify, neither block nor char device with major=0 and minor=0 can ever exist, and hence when opening such a device node will always get ENODEV, hence I think it would best to use that for overmounting device nodes. Also, maybe in order to stick to the nomenclature of stat() we should call the files we use for overmounting like this:

* /run/systemd/inaccessible/reg (empty file node of mode 000)
* /run/systemd/inaccessible/dir (empty dir of mode 000)
* /run/systemd/inaccessible/blk (block device node of major 0 and minor 0 and mode 000)
* /run/systemd/inaccessible/chr (char device node of major 0 and minor 0 and mode 000)
* /run/systemd/inaccessible/fifo (fifo node of mode 000)
* /run/systemd/inaccessible/sock (socket node of mode 000)

by naming them like this they would match the S_ISREG(), S_ISDIR(), … macros that sys/stat.h defines, neatly.